src/crystal/compiler_rt/mul.cr

```diff
@@ -0,0 +1,42 @@
+# :nodoc:
+private macro __mul_impl(name, type, n)
+  # Ported from https://github.com/llvm/llvm-project/blob/ce59ccd04023cab3a837da14079ca2dcbfebb70c/compiler-rt/lib/builtins/int_mulo_impl.inc
+  # :nodoc:
+  fun {{name}}(a : {{type}}, b : {{type}}, overflow : Int32*) : {{type}}
+    overflow.value = 0
+    result = a &* b
+    if a == {{type}}::MIN
+      if b != 0 && b != 1
+        overflow.value = 1
+      end
+      return result
+    end
+    if b == {{type}}::MIN
+      if a != 0 && a != 1
+        overflow.value = 1
+      end
+      return result
+    end
+    sa = a >> {{n - 1}}
+    abs_a = (a ^ sa) &- sa
+    sb = b >> {{n - 1}}
+    abs_b = (b ^ sb) &- sb
+    if abs_a < 2 || abs_b < 2
+      return result
+    end
+    if sa == sb
+      if abs_a > ({{type}}::MAX // abs_b)
+        overflow.value = 1
+      end
+    else
+      if abs_a > ({{type}}::MIN // -abs_b)
```

Comment:
Can we use the wrapping negative operator here like in the original Crystal implementation? (`-abs_b` is equivalent to `0 - abs_b`
```suggestion
      if abs_a > ({{type}}::MIN // ({{type}}.new(0) &- abs_b))
```

Difference should be minimal, but still.